### PR TITLE
Add query level settings: parts_to_delay_insert and parts_to_throw_insert

### LINKED ISSUE #36335: `parts_to_delay_insert` and `parts_to_throw_insert` can be also query-level settings
If they are defined, they can override table-level settings.

### PR DESCRIPTION
**Pull request by Joeywzr**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Set parts_to_delay_insert and parts_to_throw_insert as query-level settings. If they are defined, they can override table-level settings.

Closes #36335

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by alesapin** [comment]
@mergifyio update

File: src/Core/Settings.h [hunk 1]
```diff
@@ -247,6 +247,8 @@ class IColumn;
     M(Bool, fallback_to_stale_replicas_for_distributed_queries, true, "Suppose max_replica_delay_for_distributed_queries is set and all replicas for the queried table are stale. If this setting is enabled, the query will be performed anyway, otherwise the error will be reported.", 0) \
     M(UInt64, preferred_max_column_in_block_size_bytes, 0, "Limit on max column size in block while reading. Helps to decrease cache misses count. Should be close to L2 cache size.", 0) \
     \
+    M(UInt64, parts_to_delay_insert, 150, "If table contains at least that many active parts in single partition, artificially slow down insert into table.", 0) \
+    M(UInt64, parts_to_throw_insert, 300, "If more than this number active parts in single partition, throw 'Too many parts ...' exception.", 0) \
```
Comment (alesapin) [inline src/Core/Settings.h:251]:
```suggestion
    M(UInt64, parts_to_throw_insert, 300, "If more than this number active parts in a single partition of the destination table, throw 'Too many parts ...' exception.", 0) \
```

File: src/Core/Settings.h [hunk 2]
```diff
@@ -247,6 +247,8 @@ class IColumn;
     M(Bool, fallback_to_stale_replicas_for_distributed_queries, true, "Suppose max_replica_delay_for_distributed_queries is set and all replicas for the queried table are stale. If this setting is enabled, the query will be performed anyway, otherwise the error will be reported.", 0) \
     M(UInt64, preferred_max_column_in_block_size_bytes, 0, "Limit on max column size in block while reading. Helps to decrease cache misses count. Should be close to L2 cache size.", 0) \
     \
+    M(UInt64, parts_to_delay_insert, 150, "If table contains at least that many active parts in single partition, artificially slow down insert into table.", 0) \
```
Comment (alesapin) [inline src/Core/Settings.h:250]:
```suggestion
    M(UInt64, parts_to_delay_insert, 150, "If the destination table contains at least that many active parts in a single partition, artificially slow down insert into table.", 0) \
```

File: src/Storages/MergeTree/MergeTreeData.h [hunk 3]
```diff
@@ -528,7 +528,7 @@ class MergeTreeData : public IStorage, public WithMutableContext
 
     /// If the table contains too many active parts, sleep for a while to give them time to merge.
     /// If until is non-null, wake up from the sleep earlier if the event happened.
-    void delayInsertOrThrowIfNeeded(Poco::Event * until = nullptr) const;
+    void delayInsertOrThrowIfNeeded(Poco::Event * until = nullptr, ContextPtr query_context = nullptr) const;
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeData.h:533]: Do we really need a default value for arguments now? Looks like all calls of this function with two arguments now.
```suggestion
    void delayInsertOrThrowIfNeeded(Poco::Event * until, ContextPtr query_context) const;
```

**Review by alesapin — APPROVED** [review]

File: src/Storages/MergeTree/MergeTreeData.h (hunk 3 above)
Comment (Joeywzr) [reply src/Storages/MergeTree/MergeTreeData.h:533]: You are right. It can be removed.

**Comment by alesapin** [comment]
test_dns_cache -- server didn't start...

**Comment by alesapin** [comment]
Black -- fixed long time ago.

**Comment by alesapin** [comment]
Stress -- OOM, unrelated.

**Pull request merged**